### ISSUE TITLE
INNER JOIN result not showing although there are rows

### ISSUE THREAD

**Issue by datvm** [sqlitebrowser/sqlitebrowser#1892]
Details for the issue
--------------------

#### What did you do?
An INNER JOIN query does not show any result row although there should be/are. [Relevant question here](https://stackoverflow.com/questions/56115650/why-does-left-join-work-in-this-query-with-rowid-but-inner-join-does-not)

Test Schema + data:
```
CREATE TABLE OriginalValidLeaders (LeaderType TEXT);

INSERT INTO OriginalValidLeaders (LeaderType) VALUES
('Leader A'), ('Leader B'), ('Leader C');
```

Query (I want to assign each Leader with a random Leader):

```
SELECT L.LeaderType, R.LeaderType
FROM OriginalValidLeaders L
INNER JOIN
    (
        SELECT LeaderType
        FROM OriginalValidLeaders
        ORDER BY random()
    ) R
ON L.rowid = R.rowid;
```

#### What did you expect to see?
3 rows of result.

#### What did you see instead?
No result shown, although SELECT COUNT(*) returns 3 (and the UI also shows 3).

Useful extra information
-------------------------
If I use LEFT JOIN instead, the result shows as expected, but I do not understand why LEFT JOIN works but not INNER JOIN.

When I tried the Nightly build, it shows 3 row result but no column.

#### What operating system are you using?

- [x] Windows: ( _version: 10 )
- [ ] Linux: ( _distro:_ ___ )
- [ ] Mac OS: ( _version:_ ___ )
- [ ] Other: ___

#### What is your DB4S version?

- [ ] 3.11.2
- [ ] 3.11.1
- [x] 3.10.1
- [ ] Other: ___

#### Did you also

- [x] Try out the latest nightly version: https://github.com/sqlitebrowser/sqlitebrowser#nightly-builds
- [x] Search for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by justinclift** [comment]
Hmmm, if I'm reading that correctly and it shows 3 rows in the result but doesn't actually display the columns... that sounds like a bug in our code for displaying output.

Any chance you could screenshot it, just to make it clear? :smile:

**Comment by datvm** [comment]
Sorry I intended to add screenshot but was busy with something else. Sure it would clarify a lot. Here it is:

![image](https://user-images.githubusercontent.com/6388546/57711085-4bc7c280-7698-11e9-8c55-571f67aeab22.png)


![image](https://user-images.githubusercontent.com/6388546/57711191-7ade3400-7698-11e9-8d4e-7faad1c05dca.png)

**Comment by datvm** [comment]
Hey I have just discovered a new weird one: on 3.10.1, if I change it to LEFT JOIN, I get the result I desire:

![image](https://user-images.githubusercontent.com/6388546/57711311-acef9600-7698-11e9-8bc3-4e8ff3ec640e.png)

But on nightly build, LEFT JOIN returns nothing for 2nd column:

![image](https://user-images.githubusercontent.com/6388546/57711353-c0026600-7698-11e9-9f15-02a679b95ae8.png)

**Comment by justinclift** [comment]
Thanks @datvm.  Yeah, not sure.  The result clearly says it's returning rows though, so it's probably something we should look at just to be on the safe side. :smile:

**Comment by scottfurry** [comment]
I would be curious to see the underlying database, if you can share. There's something to that query I would like to look at.

On another note, the notion of expanding usage of QTest to database handling and JOINS is raised. This may be a interesting exercise for future consideration.

**Comment by datvm** [comment]
The issue itself can be fully created using the above schema. If you meant you were curious about the game database, then sure, it's from the game Civ 6:

[DebugGameplay.zip](https://github.com/sqlitebrowser/sqlitebrowser/files/3180384/DebugGameplay.zip)

**Comment by scottfurry** [comment]
That's what I thought...
```
SELECT L.LeaderType, R.LeaderType
FROM OriginalValidLeaders L, OriginalValidLeaders R
INNER JOIN(SELECT LeaderType FROM OriginalValidLeaders ORDER BY random())
ON L.rowid = R.RowId;
```
Gives full values right and left.
![DB4S_CIV6_JOINS](https://user-images.githubusercontent.com/6686571/57740791-bff16e80-7676-11e9-9173-72999818fe58.png)

There was something about the alias being used after the select sub query that I needed to see for myself. The data being fully presented when you tried LEFT JOIN *sort of* makes sense to me as the Left column is defined in the FROM clause.

However, results are being returned with the INNER JOIN statement you used - just not presented. My suspicion is that an unqualified reference to column causes some kind of data presentation glitch. To me this is some kind of bug. Some involved debugging is going to be needed here. Something, IMHO, should be sorted out. I'm still trying to understand *how* the INNER JOIN you used was able to work. (Then again JOINS are to SQL people as regex is to some shell/programming people - some kind of voodoo magic that mysteriously works for a specific solution every now and again.)

To complicate matters, when I tried using the ORDER BY change, from your earlier post, with my tweak to FROM clause as so:
```
SELECT L.LeaderType, R.LeaderType
FROM OriginalValidLeaders L, OriginalValidLeaders R
INNER JOIN(SELECT LeaderType FROM OriginalValidLeaders ORDER BY random())
ON L.rowid = R.RowId
ORDER BY L.rowid;
```
I got a table full of the same randomly chosen leader. So, yeah. Something is not right here.

To me, this is definitely prime territory for some kind of testing. Framework in my head is something along the lines of:
a) Test database
b) Test case to execute SQL on test database
c) read resultant table widget
...
And pushing boundaries on JOINS would be useful, if not outright reassuring. This is going to be a longer term effort.

**Comment by datvm** [comment]
Glad I could contribute. Good luck with it. I am just a developer and not database expert so I guess it's all magic for me from this point.

P.s: there seem to be inconsistency between online SQL Fiddlers as well, so it might actually be problems in connector/engine (or whatever you guys call in SQLite).

**Comment by MKleusberg** [comment]
This definitely looks like a problem in SQLite. While playing around with different workarounds, I e.g. noticed that this is working fine, even though I don't see why it should make a difference:
```sql
SELECT L.LeaderType, R.LeaderType
FROM OriginalValidLeaders L
INNER JOIN
    (
        SELECT row_number() over (order by random()) as n, LeaderType
        FROM OriginalValidLeaders
        ORDER BY random()
    ) R
ON L.rowid = R.rowid
```

As for the display problem in the GUI: That's because SQLite actually returns some rows here but those rows have 0 columns. And that's exactly what you see in the grid view. I would argue that SQLite should report this as 0 rows in the first place because arguably no columns means no data. Any opinions on whether we should work around this problem and show 0 rows? It shouldn't be too hard to add but I'm not sure if this solving or hiding the problem :wink:

**Comment by justinclift** [comment]
Yeah, not sure.  Maybe it's something we should ask the SQLite dev's about, as showing a row count with no columns seems like it might need adjust there?

**Comment by mlawren** [comment]
I cannot speak to the SQLiteBrowser display issues but your subquery "R" does not actually define a rowid. So I suspect that SQLite is filling that in automatically with NULLs.

If you are explicit about them comming from the original table your query works:

```
SELECT L.LeaderType, R.LeaderType
FROM OriginalValidLeaders L
INNER JOIN
    (
        SELECT rowid, LeaderType  -- Added a rowid column here
        FROM OriginalValidLeaders
        ORDER BY random()
    ) R
ON L.rowid = R.rowid;
```
